core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java

```diff
@@ -640,26 +617,22 @@ public boolean equals(Object o) {
         if (this == o) {
             return true;
         }
-        // we check on instanceof so we also handle the ImmutableShardRouting case as well
         if (o == null || !(o instanceof ShardRouting)) {
             return false;
         }
         ShardRouting that = (ShardRouting) o;
-        if (version != that.version) {
-            return false;
-        }
         if (unassignedInfo != null ? !unassignedInfo.equals(that.unassignedInfo) : that.unassignedInfo != null) {
             return false;
         }
         return equalsIgnoringMetaData(that);
     }
 
-    private long hashVersion = version - 1;
+    private boolean usePreComputedHashCode = false;
     private int hashCode = 0;
 
     @Override
     public int hashCode() {
-        if (hashVersion == version) {
+        if (frozen && usePreComputedHashCode) {
```

Comment:
I see, basically the hashVersion was never set. I'm +1 on the fix here.

> there is one where non-frozen ShardRouting is used as map key and that is in BalancedShardsAllocator.

Yeah, that sounds like a trouble waiting to happen. Agree we should investigate and propose a fix in another change.